### ISSUE TITLE
Difference between `[float] * n` and `[float for _ in range(n)]` for type checking?

### ISSUE THREAD

**Issue by alechouse97** [astral-sh/ty#3013]
### Question

I have this test script:

```python
def func1(x: list[type] | None = None):
    y: list[type] = x if x is not None else [float for _ in range(3)]

def func2(x: list[type] | None = None):
    y: list[type] = x if x is not None else [float] * 3
```

`func1` passes the checker without issue, but `func2` receives the error:

```
error[invalid-assignment]: Object of type `list[type] | list[<class 'float'>]` is not assignable to `list[type]`
 --> tmp.py:6:8
  |
5 | def func2(x: list[type] | None = None):
6 |     y: list[type] = x if x is not None else [float] * 3
  |        ----------   ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ Incompatible value of type `list[type] | list[<class 'float'>]`
  |        |
  |        Declared type
```

In terms of type checking, what's the difference between `[float] * 2` and `[float for _ in range(2)]`

### Version

ty 0.0.21

**Comment by sharkdp** [comment]
Thank you for reporting this. A slightly more minified example would be
```py
y1: list[type] = [float for _ in range(3)]
y2: list[type] = [float] * 3  # `list[<class 'float'>]` is not assignable to `list[type]`
```

The problem is that the type context (`list[type]`) is not propagated through the binary operator type inference (the type inference for the implicit `list.__mul__` call). This means we infer `[float]` without type context as `list[<class 'float'>]`, which is not assignable to `list[type]` due to invariance of `list`.

@ibraheemdev FYI. This seems similar to what you mentioned here: https://github.com/astral-sh/ty/issues/2986#issuecomment-4017872843, but might be quite tricky to support(?).

**Comment by sharkdp** [comment]
I think https://github.com/astral-sh/ty/issues/2972 would also help here, because `<class 'float'>` is a singleton type.

**Comment by carljm** [comment]
I think the "pass type context through multiplication" issue here is exactly the same as in https://github.com/astral-sh/ty/issues/3002.

Fixing this solely via literal promotion (without propagating type context) is tricky, because it's less clear what `<class 'float'>` should promote to than it is for, say, `Literal[0]`. To help this particular example, you'd need to promote all the way to `type` (not just to `type[float]`). But is that always desirable? Are there other use cases where that would be over-promoting?

**Comment by ibraheemdev** [comment]
This was also resolved by https://github.com/astral-sh/ruff/pull/24197.
